python/lsst/ts/salobj/configurable_csc.py

```diff
@@ -0,0 +1,151 @@
+# This product includes software developed by the LSST Project
+# (https://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+__all__ = ["ConfigurableCsc"]
+
+import abc
+import os
+import types
+
+import yaml
+
+from . import base
+from .base_csc import BaseCsc, State
+from .validator import DefaultingValidator
+
+
+class ConfigurableCsc(BaseCsc, abc.ABC):
+    """Base class for a configurable Commandable SAL Component (CSC)
+
+    Parameters
+    ----------
+    sallib : ``module``
+        salpy component library generatedby SAL
+    index : `int` or `None`
+        SAL component index, or 0 or None if the component is not indexed.
+    config_dir : `str`
+        Path to configuration files.
+    schema_path : `str`
+        Path to a schema file used to validate configuration files.
+    initial_state : `State` or `int` (optional)
+        The initial state of the CSC. This is provided for unit testing,
+        as real CSCs should start up in `State.STANDBY`, the default.
+    initial_simulation_mode : `int` (optional)
+        Initial simulation mode. This is provided for unit testing,
+        as real CSCs should start up not simulating, the default.
+
+    Raises
+    ------
+    ValueError
+        If ``config_dir`` is not a directory.
+    salobj.ExpectedError
+        If ``initial_state`` or ``initial_simulation_mode`` is invalid.
+
+    Notes
+    -----
+    Configuration is handled by the ``start`` command, as follows:
+
+    * The ``settingsToApply`` field specifies a path to a configuration file
+      relative to ``config_dir``.
+    * The configuration file is validated against the schema specified
+      by specified ``schema_path``.
+      This includes setting default values from the schema and validating
+      the result again (in case the default values are invalid).
+    * The validated configuration is converted to a struct-like object
+      using `types.SimpleNamespace`.
+    * The configuration is passed to the ``configure`` method,
+      which subclasses must override. Note that ``configure`` is called just
+      before summary state changes from `State.STANDBY` to `State.DISABLED`.
+    """
+    def __init__(self, sallib, index, config_dir, schema_path,
```

Comment:
CSCs should publish a `settingVersions` event with the `recommendedSettingsVersion` and `recommendedSettingsLabels`. 

I think that a Configurable CSC has enough information to publish this event so I think it would be great to do that by default. One caveat is that it must be re-published anytime the CSC goes into standby.

The `recommendedSettingsLabels` contains all available settings and `recommendedSettingsVersion` the version of the configuration. 

This brings the question as to how we manage the version of a configuration. I think that right now we could use the `__fingerprint__` in `version.py`.

Reply:
Where are these documented? I see them in SALGenerics.xml but they have no description. I'm especially puzzled by `recommendedSettingsVersion` since the only thing the `start` command takes is a single string which is the name of the config file. It seems to me that users need a list of the names that can be provided to `start`. From this I conclude:

• `recommendedSettingsLabels` should be a list of files in the config directory
• If the config directory contains multiple versions of a given config then they need to be differentiated by file name (easily done)

Another worry is that `recommendedSettingsLabels` is only 256 chars long. So it could easily be too short to list all available configuration files, especially if there is a hierarchy.

Reply:
That is a very good point. I'm looking for the documentation regarding this and will let you know about my findings. What I know is from meetings with SE team that led the effort on developing that topic.

Reply:
Thank you. I'm glad you caught this omission and I clearly need to add support for it. I'm just not sure how to go about it.

Reply:
We can probably leave it as it is for now and deal with that later once we have a better understanding of the solution.

Reply:
I just talked to @CContaxis about this, since he had to deal with it for M1M3. To summarize what he does (I've copied him on this so he can correct any errors):
- `recommendedSettingsLabels` contains a list of the recommended settings labels. It does not include other versions that are not recommended.
- Old versions are always retained.
- M1M3 uses an alias so the real files always have a version name and the alias probably does not. This has the advantage that you can always use the same string to mean the same basic configuration, but it has the disadvantage that you cannot tell which file was actually used. I think if we go that route then we must use the `recommendedSettingsVersion` field to identify the git version or something. Still...I prefer explicit to implicit.

My own inclination is to make the entries in `recommendedSettingsLabels` include include version information (and not use aliases). I am less ure what to put in `recommendedSettingsVersion` but I also doubt it matters very much.